### ISSUE TITLE
Flooded logging when SQL connectivity is intermitant 

### ISSUE THREAD

**Issue by SimonCropp** [Particular/NServiceBus.SqlServer#654]
During SQL patching we occasionally get a flood of logging with  2  errors 

## SQL Peek operation failed

```
System.Transactions.TransactionInDoubtException: The transaction is in doubt. ---> System.Data.SqlClient.SqlException: Execution Timeout Expired. The timeout period elapsed prior to completion of the operation or the server is not responding. ---> System.ComponentModel.Win32Exception: The wait operation timed out
--- End of inner exception stack trace --- 
at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction) 
at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose) 
at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady) 
at System.Data.SqlClient.TdsParser.Run(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj) 
at System.Data.SqlClient.TdsParser.TdsExecuteTransactionManagerRequest(Byte[] buffer, TransactionManagerRequestType request, String transactionName, TransactionManagerIsolationLevel isoLevel, Int32 timeout, SqlInternalTransaction transaction, TdsParserStateObject stateObj, Boolean isDelegateControlRequest) 
at System.Data.SqlClient.SqlInternalConnectionTds.ExecuteTransactionYukon(TransactionRequest transactionRequest, String transactionName, IsolationLevel iso, SqlInternalTransaction internalTransaction, Boolean isDelegateControlRequest) at System.Data.SqlClient.SqlDelegatedTransaction.SinglePhaseCommit(SinglePhaseEnlistment enlistment)
--- End of inner exception stack trace --- 
at System.Transactions.TransactionStateInDoubt.EndCommit(InternalTransaction tx) 
at System.Transactions.CommittableTransaction.Commit() 
at System.Transactions.TransactionScope.InternalDispose() 
at System.Transactions.TransactionScope.Dispose() 
at NServiceBus.Transport.SQLServer.QueuePeeker.<Peek>d__1.MoveNext()
```

## Exception thrown while performing cancellation
```
System.Data.SqlClient.SqlException (0x80131904): Connection Timeout Expired.  The timeout period elapsed while attempting to consume the pre-login handshake acknowledgement.  This could be because the pre-login handshake failed or the server was unable to respond back in time.  The duration spent while attempting to connect to this server was - [Pre-Login] initialization=10220; handshake=4790;  ---> System.ComponentModel.Win32Exception (258): The wait operation timed out   
at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)   
at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal& connection)\r\n   at System.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)   
at System.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)\r\n   
at System.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry)   
at System.Data.SqlClient.SqlConnection.OpenAsync(CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---   
at NServiceBus.Transport.SQLServer.SqlConnectionFactory.<>c__DisplayClass2_0.<<Default>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---   
at NServiceBus.Transport.SQLServer.SqlConnectionFactory.OpenNewConnection()   
at NServiceBus.Transport.SQLServer.DelayedMessageHandler.MoveMaturedDelayedMessages()
```

## State

net48

```
    <PackageReference Include="NServiceBus" Version="7.2.0" />
    <PackageReference Include="NServiceBus.Newtonsoft.Json" Version="2.2.0" />
    <PackageReference Include="NServiceBus.Persistence.Sql" Version="4.6.1" />
    <PackageReference Include="NServiceBus.SqlServer" Version="4.3.1" />
```
I know we are using an older versions, but i could not see anything in the recent release notes that would address this?

## Critical error

is currently

```
	string fatalMessage = "The following critical error was encountered:\n" + context.get_Error() + "\nProcess is shutting down.";
		Log.Fatal(context.get_Exception(), fatalMessage);
		Log.CloseAndFlush();
		Environment.FailFast(fatalMessage, context.get_Exception());
		return Task.CompletedTask;
```

and when u run locally, and stop sql while running, the critical error is correctly hit and the endpoint shits down

![SQL Peek operation failed](https://user-images.githubusercontent.com/122666/91146719-9e48b900-e6fa-11ea-9c2d-036aa16d57be.PNG)
![Exception thrown while performing cancellation](https://user-images.githubusercontent.com/122666/91146725-9f79e600-e6fa-11ea-9404-5894114c8d6f.PNG)

**Comment by SimonCropp** [comment]
Speculating  here...

This is current code of QueuePeeker.Peek

```
public async Task<int> Peek(TableBasedQueue inputQueue, RepeatedFailuresOverTimeCircuitBreaker circuitBreaker, CancellationToken cancellationToken)
{
    var messageCount = 0;

    try
    {
        using (var scope = new TransactionScope(TransactionScopeOption.RequiresNew, new TransactionOptions { IsolationLevel = IsolationLevel.ReadCommitted }, TransactionScopeAsyncFlowOption.Enabled))
        using (var connection = await connectionFactory.OpenNewConnection().ConfigureAwait(false))
        {
            messageCount = await inputQueue.TryPeek(connection, null, cancellationToken).ConfigureAwait(false);

            circuitBreaker.Success();

            if (messageCount == 0)
            {
                Logger.Debug($"Input queue empty. Next peek operation will be delayed for {settings.Delay}.");

                await Task.Delay(settings.Delay, cancellationToken).ConfigureAwait(false);
            }

            scope.Complete();
        }
    }
    catch (OperationCanceledException)
    {
        //Graceful shutdown
    }
    catch (SqlException e) when (cancellationToken.IsCancellationRequested)
    {
        Logger.Debug("Exception thrown while performing cancellation", e);
    }
    catch (Exception ex)
    {
        Logger.Warn("Sql peek operation failed", ex);
        await circuitBreaker.Failure(ex).ConfigureAwait(false);
    }

    return messageCount;
}
```

So if there are not messages, `inputQueue.TryPeek` succeeds and messageCount==0. then the transaction fails. depending on the exact exception either 

```
Logger.Debug("Exception thrown while performing cancellation", e);
```

or 

```
Logger.Warn("Sql peek operation failed", ex);
await circuitBreaker.Failure(ex).ConfigureAwait(false);
```

will be hit. but nothing will bubble

Peek is called from 

```
async Task InnerProcessMessages()
{
    while (!cancellationTokenSource.IsCancellationRequested)
    {
        var messageCount = await queuePeeker.Peek(inputQueue, peekCircuitBreaker, cancellationToken).ConfigureAwait(false);

        if (messageCount == 0)
        {
            continue;
        }

        if (cancellationTokenSource.IsCancellationRequested)
        {
            return;
        }

        // We cannot dispose this token source because of potential race conditions of concurrent receives
        var loopCancellationTokenSource = new CancellationTokenSource();

        // If the receive circuit breaker is triggered start only one message processing task at a time.
        var maximumConcurrentReceives = receiveCircuitBreaker.Triggered ? 1 : messageCount;

        for (var i = 0; i < maximumConcurrentReceives; i++)
        {
            if (loopCancellationTokenSource.Token.IsCancellationRequested)
            {
                break;
            }

            await concurrencyLimiter.WaitAsync(cancellationToken).ConfigureAwait(false);

            var receiveTask = InnerReceive(loopCancellationTokenSource);
            runningReceiveTasks.TryAdd(receiveTask, receiveTask);

            receiveTask.ContinueWith((t, state) =>
            {
                var receiveTasks = (ConcurrentDictionary<Task, Task>)state;
                receiveTasks.TryRemove(t, out _);
            }, runningReceiveTasks, TaskContinuationOptions.ExecuteSynchronously)
            .Ignore();
        }
    }
}
```

So the exception wont bubble. and 

```
 if (messageCount == 0)
{
    continue;
}
```
results in a fast loop?

**Comment by SimonCropp** [comment]
Speculating  here...

This is current code of QueuePeeker.Peek

```
public async Task<int> Peek(TableBasedQueue inputQueue, RepeatedFailuresOverTimeCircuitBreaker circuitBreaker, CancellationToken cancellationToken)
{
    var messageCount = 0;

    try
    {
        using (var scope = new TransactionScope(TransactionScopeOption.RequiresNew, new TransactionOptions { IsolationLevel = IsolationLevel.ReadCommitted }, TransactionScopeAsyncFlowOption.Enabled))
        using (var connection = await connectionFactory.OpenNewConnection().ConfigureAwait(false))
        {
            messageCount = await inputQueue.TryPeek(connection, null, cancellationToken).ConfigureAwait(false);

            circuitBreaker.Success();

            if (messageCount == 0)
            {
                Logger.Debug($"Input queue empty. Next peek operation will be delayed for {settings.Delay}.");

                await Task.Delay(settings.Delay, cancellationToken).ConfigureAwait(false);
            }

            scope.Complete();
        }
    }
    catch (OperationCanceledException)
    {
        //Graceful shutdown
    }
    catch (SqlException e) when (cancellationToken.IsCancellationRequested)
    {
        Logger.Debug("Exception thrown while performing cancellation", e);
    }
    catch (Exception ex)
    {
        Logger.Warn("Sql peek operation failed", ex);
        await circuitBreaker.Failure(ex).ConfigureAwait(false);
    }

    return messageCount;
}
```

So if there are messages, `inputQueue.TryPeek` succeeds, messageCount!=0, and `circuitBreaker.Success`. then the transaction fails with TransactionInDoubtException (as per the above stack trace). depending on the exact exception either 

```
Logger.Warn("Sql peek operation failed", ex);
await circuitBreaker.Failure(ex).ConfigureAwait(false);
```

so doesnt this result in a circuitBreaker being constantly armed and disarmed?

**Comment by SimonCropp** [comment]
Not familiar with the code. but i would have assumed peek should look like this

```
public async Task<int> Peek(TableBasedQueue inputQueue, RepeatedFailuresOverTimeCircuitBreaker circuitBreaker, CancellationToken cancellationToken)
{
    var messageCount = 0;

    try
    {
        using (var scope = new TransactionScope(TransactionScopeOption.RequiresNew, new TransactionOptions { IsolationLevel = IsolationLevel.ReadCommitted }, TransactionScopeAsyncFlowOption.Enabled))
        using (var connection = await connectionFactory.OpenNewConnection().ConfigureAwait(false))
        {
            messageCount = await inputQueue.TryPeek(connection, null, cancellationToken).ConfigureAwait(false);
            scope.Complete();
        }
        circuitBreaker.Success();

        if (messageCount == 0)
        {
            Logger.Debug($"Input queue empty. Next peek operation will be delayed for {settings.Delay}.");

            await Task.Delay(settings.Delay, cancellationToken).ConfigureAwait(false);
        }
    }
    catch (OperationCanceledException)
    {
        //Graceful shutdown
    }
    catch (SqlException e) when (cancellationToken.IsCancellationRequested)
    {
        Logger.Debug("Exception thrown while performing cancellation", e);
    }
    catch (Exception ex)
    {
        Logger.Warn("Sql peek operation failed", ex);
        await circuitBreaker.Failure(ex).ConfigureAwait(false);
    }

    return messageCount;
}
```

**Comment by danielmarbach** [comment]
@SimonCropp thanks for the detailed bug report. You are spot on with your proposed changes because the scope is currently held open unnecessarily long which can increase the chance of running into `TransactionInDoubtException`. We will look more into this and keep you posted

**Comment by danielmarbach** [comment]
@SimonCropp I've raised a [PR](https://github.com/Particular/NServiceBus.SqlServer/pull/660)

The PR moves the delay out of the scope completion to not unnaturally delay the transaction from being completed in the case no messages are returned. From the perspective of the TransactionInDoubt due to timeouts or deadlocks while other operations are going on (like patching the DB) we still feel the right choice is to trigger the circuit breaker. Right now we are not triggering the CB when shutting down in case of SQL exceptions like you mentioned but triggering the CB for all exceptions during normal operations. 

You may want to consider changing the wait time?

https://docs.particular.net/transports/sql/connection-settings#circuit-breaker-wait-time

by the way, the circuit breaker already does a delay in case of a failure.

But maybe I misunderstood the history of this issue. Did you get the TransactionInDoubt also during a shutdown?

**Comment by SimonCropp** [comment]
> You may want to consider changing the wait time?

I dont think that would help. from my comment above:

> so doesnt this result in a circuitBreaker being constantly armed and disarmed?

so i dont think the circuit breaker fires?

> Did you get the TransactionInDoubt also during a shutdown?

from my analysis it is not during shutdown

**Comment by danielmarbach** [comment]
Fixes will be released next week as part of 6.1.2

**Comment by SimonCropp** [comment]
@danielmarbach thanks

**Comment by danielmarbach** [comment]
[We have released 6.1.2](https://discuss.particular.net/t/nservicebus-sqlserver-6-1-2-patch-release-available/2058)
